crates/data-transformer/src/shared/path.rs

```diff
@@ -0,0 +1,85 @@
+use cairo_lang_parser::utils::SimpleParserDatabase;
+use cairo_lang_syntax::node::ast::{
+    ExprPath, GenericArg, GenericArgValue, PathSegment, PathSegmentWithGenericArgs,
+};
+use cairo_lang_syntax::node::{Token, TypedSyntaxNode};
+
+#[derive(Debug, thiserror::Error)]
+pub enum GenericArgsExtractionError {
+    #[error("Invalid generic arguments")]
+    InvalidGenericArgs,
+    #[error("Expected exactly one generic argument")]
+    MoreThanOneGenericArg,
+}
+
+pub enum SplitResult {
+    Simple {
+        splits: Vec<String>,
+    },
+    WithGenericArgs {
+        splits: Vec<String>,
+        generic_args: String,
+    },
+}
+
+/// Splits a path into its segments, and extracts generic arguments if present.
+///
+/// In the case of Cairo-like language constructs such as arrays or spans,
+/// we assume that if there are generic arguments (e.g., `Span<T>`), they
+/// appear at the end of the path. Therefore, by the time we encounter a
+/// segment with generic arguments, all preceding segments have already
+/// been collected.
+///
+/// For example, in a path like `core::array::Array<felt252>`, this function will:
+/// - Collect "core", "array", and "Array" into `splits`
+/// - Extract the generic argument `felt252` from `Array<felt252>`
+pub fn split(
+    path: &ExprPath,
+    db: &SimpleParserDatabase,
+) -> Result<SplitResult, GenericArgsExtractionError> {
+    let mut splits = Vec::new();
+    for p in path.elements(db) {
+        match p {
+            PathSegment::Simple(segment) => {
+                splits.push(segment.ident(db).token(db).text(db).to_string());
+            }
+            PathSegment::WithGenericArgs(segment) => {
+                splits.push(segment.ident(db).token(db).text(db).to_string());
+                let generic_args = extract_generic_args(&segment, db)?;
+                return Ok(SplitResult::WithGenericArgs {
+                    splits,
+                    generic_args,
+                });
```

Comment:
Maybe there should be a safeguard assert here that checks if the iterator has been exhausted before we return? In case of malformed abi.

Or is it verified in some other step?